File: packages/js/node/test/durable-revocation.test.js
import assert from "node:assert/strict";
import { generateKeyPairSync } from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";

import { ArtifactSignatureError, SecurityError } from "@handoffkit/csp";
import {
  ArtifactSigner,
  ArtifactSigningCredential,
  ArtifactTrustPolicy,
  DurableRevocationPolicy,
  RevocationEntry,
  verifySignedArtifact,
} from "../src/index.js";

const sharedFixture = fileURLToPath(
  new URL("../../../contracts/test-fixtures/security/durable-revocation-v1.json", import.meta.url),
);

function workspace(t) {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), "handoffkit-node-revocation-"));
  t.after(() => fs.rmSync(root, { recursive: true, force: true }));
  return root;
}

function entry(kind = "certificate_fingerprint", value = `sha256:${"a".repeat(64)}`, options = {}) {
  const revokedAt = options.revokedAt ?? Math.floor(Date.now() / 1000);
  return new RevocationEntry({
    kind,
    value,
    reason: "credential compromise",
    revokedAt,
    effectiveAt: options.effectiveAt,
    expiresAt: options.expiresAt,
  });
}

test("Node durable revocation persists all supported subjects", (t) => {
  const filePath = path.join(workspace(t), "revocations.json");
  const policy = new DurableRevocationPolicy(filePath);
  policy.revoke(entry());
  policy.revoke(entry("peer_id", "peer-a"));
  policy.revoke(entry("issuer", "CN=HandoffKit Test CA"));
  policy.revoke(entry("trust_domain", "HANDOFFKIT.INTERNAL"));

  const restored = new DurableRevocationPolicy(filePath);
  assert.equal(restored.status().generation, 4);
  assert.equal(restored.isRevoked("certificate_fingerprint", `${"AA:".repeat(31)}AA`), true);
  assert.equal(restored.isRevoked("peer_id", "peer-a"), true);
  assert.equal(restored.isRevoked("issuer", "CN=HandoffKit Test CA"), true);
  assert.equal(restored.isRevoked("trust_domain", "handoffkit.internal"), true);
  assert.equal(restored.isRevoked("peer_id", "peer-b"), false);
  assert.equal(restored.isRevoked("certificate_fingerprint", `sha256:${"b".repeat(64)}`), false);
});

test("Node loads the shared durable revocation fixture", (t) => {
  const filePath = path.join(workspace(t), "shared-revocations.json");
  fs.copyFileSync(sharedFixture, filePath);
  if (process.platform !== "win32") fs.chmodSync(filePath, 0o600);
  const policy = new DurableRevocationPolicy(filePath);
  assert.deepEqual(policy.status({ now: 1_800_000_000 }), {
    format: "handoffkit.security.revocations",
    format_version: 1,
    generation: 3,
    entries: 3,
    active: 2,
  });
  assert.equal(policy.isRevoked(
    "certificate_fingerprint",
    `sha256:${"a".repeat(64)}`,
    { now: 1_800_000_000 },
  ), true);
  assert.equal(policy.isRevoked(
    "signer_fingerprint",
    `sha256:${"b".repeat(64)}`,
    { now: 1_800_000_000 },
  ), true);
  assert.equal(policy.isRevoked("peer_id", "peer-b", { now: 1_800_000_000 }), false);
});

test("Node durable revocation supports effective windows, remove, and live reload", (t) => {
  const filePath = path.join(workspace(t), "revocations.json");
  const reader = new DurableRevocationPolicy(filePath);
  const writer = new DurableRevocationPolicy(filePath);
  const now = Math.floor(Date.now() / 1000);
  writer.revoke(entry("peer_id", "future-peer", {
    revokedAt: now,
    effectiveAt: now + 10,
    expiresAt: now + 20,
  }));
  assert.equal(reader.isRevoked("peer_id", "future-peer", { now: now + 11 }), false);
  reader.reload();
  assert.equal(reader.isRevoked("peer_id", "future-peer", { now: now + 9 }), false);
  assert.equal(reader.isRevoked("peer_id", "future-peer", { now: now + 10 }), true);
  assert.equal(reader.isRevoked("peer_id", "future-peer", { now: now + 20 }), false);
  assert.equal(writer.remove("peer_id", "future-peer"), true);
  reader.reload();
  assert.equal(reader.isRevoked("peer_id", "future-peer", { now: now + 11 }), false);
});

test("Node revocation capacity fails closed and corruption is quarantined", (t) => {
  const root = workspace(t);
  const filePath = path.join(root, "revocations.json");
  const policy = new DurableRevocationPolicy(filePath, { maxEntries: 1 });
  policy.revoke(entry());
  assert.throws(
    () => policy.revoke(entry("peer_id", "peer-a")),
    (error) => error instanceof SecurityError && error.code === "revocation_state_capacity",
  );
  const value = JSON.parse(fs.readFileSync(filePath, "utf8"));
  value.checksum = "sha256:00";
  fs.writeFileSync(filePath, JSON.stringify(value), { mode: 0o600 });
  if (process.platform !== "win32") fs.chmodSync(filePath, 0o600);
  assert.throws(
    () => new DurableRevocationPolicy(filePath),
    (error) => error instanceof SecurityError && error.code === "security_state_corrupt",
  );
  assert.equal(fs.existsSync(filePath), false);
  assert.ok(fs.readdirSync(root).some((name) => name.includes("corrupt-")));
});

test("Node Ed25519 verification enforces durable signer revocation", (t) => {
  const { privateKey, publicKey } = generateKeyPairSync("ed25519");
  const credential = new ArtifactSigningCredential({
    signerIdentity: "producer-a",
    publicKey: publicKey.export({ type: "spki", format: "pem" }),
  });
  const signer = new ArtifactSigner(
    privateKey.export({ type: "pkcs8", format: "pem" }),
    "producer-a",
  );
  const artifact = signer.sign("artifact-a", Buffer.from("verified payload"));
  const revocations = new DurableRevocationPolicy(
    path.join(workspace(t), "revocations.json"),
  );
  const policy = new ArtifactTrustPolicy([credential], { revocationPolicy: revocations });
  assert.equal(verifySignedArtifact(Buffer.from("verified payload"), artifact, policy), true);
  revocations.revoke(entry("signer_fingerprint", credential.fingerprint));
  assert.throws(
    () => verifySignedArtifact(Buffer.from("verified payload"), artifact, policy),
    (error) => error instanceof ArtifactSignatureError && error.code === "artifact_signer_revoked",
  );
});

File: packages/go/security/durable_revocation_test.go
package security

import (
	"crypto/ed25519"
	"crypto/rand"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"
)

func revocationEntry(t *testing.T, kind RevocationKind, value string, now, effectiveAt, expiresAt int64) RevocationEntry {
	t.Helper()
	entry, err := NewRevocationEntry(kind, value, "credential compromise", now, effectiveAt, expiresAt)
	if err != nil {
		t.Fatal(err)
	}
	return entry
}

func revocationOptions() DurableRevocationOptions {
	return DefaultDurableRevocationOptions()
}

func TestDurableRevocationPersistsAndScopesSubjects(t *testing.T) {
	path := filepath.Join(t.TempDir(), "revocations.json")
	policy, err := NewDurableRevocationPolicy(path, revocationOptions())
	if err != nil {
		t.Fatal(err)
	}
	now := time.Now().Unix()
	for _, entry := range []RevocationEntry{
		revocationEntry(t, RevocationCertificateFingerprint, "sha256:"+strings.Repeat("a", 64), now, 0, 0),
		revocationEntry(t, RevocationPeerID, "peer-a", now, 0, 0),
		revocationEntry(t, RevocationIssuer, "CN=HandoffKit Test CA", now, 0, 0),
		revocationEntry(t, RevocationTrustDomain, "HANDOFFKIT.INTERNAL", now, 0, 0),
	} {
		if err := policy.Revoke(entry); err != nil {
			t.Fatal(err)
		}
	}
	restored, err := NewDurableRevocationPolicy(path, revocationOptions())
	if err != nil {
		t.Fatal(err)
	}
	if status := restored.Status(now); status.Generation != 4 || status.Entries != 4 || status.Active != 4 {
		t.Fatalf("unexpected revocation status: %#v", status)
	}
	for _, candidate := range []struct {
		kind  RevocationKind
		value string
	}{
		{RevocationCertificateFingerprint, strings.Repeat("AA:", 31) + "AA"},
		{RevocationPeerID, "peer-a"},
		{RevocationIssuer, "CN=HandoffKit Test CA"},
		{RevocationTrustDomain, "handoffkit.internal"},
	} {
		revoked, checkErr := restored.IsRevoked(candidate.kind, candidate.value, now)
		if checkErr != nil || !revoked {
			t.Fatalf("expected %s revocation, got %v: %v", candidate.kind, revoked, checkErr)
		}
	}
	if revoked, _ := restored.IsRevoked(RevocationCertificateFingerprint, "sha256:"+strings.Repeat("b", 64), now); revoked {
		t.Fatal("renewed credential was incorrectly revoked")
	}
}

func TestGoLoadsSharedDurableRevocationFixture(t *testing.T) {
	_, filename, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("cannot locate durable revocation test")
	}
	fixture := filepath.Clean(filepath.Join(filepath.Dir(filename), "..", "..", "contracts", "test-fixtures", "security", "durable-revocation-v1.json"))
	raw, err := os.ReadFile(fixture)
	if err != nil {
		t.Fatal(err)
	}
	path := filepath.Join(t.TempDir(), "shared-revocations.json")
	if err := os.WriteFile(path, raw, 0o600); err != nil {
		t.Fatal(err)
	}
	policy, err := NewDurableRevocationPolicy(path, revocationOptions())
	if err != nil {
		t.Fatal(err)
	}
	status := policy.Status(1_800_000_000)
	if status.Generation != 3 || status.Entries != 3 || status.Active != 2 {
		t.Fatalf("shared revocation status mismatch: %#v", status)
	}
	for _, candidate := range []struct {
		kind     RevocationKind
		value    string
		expected bool
	}{
		{RevocationCertificateFingerprint, "sha256:" + strings.Repeat("a", 64), true},
		{RevocationSignerFingerprint, "sha256:" + strings.Repeat("b", 64), true},
		{RevocationPeerID, "peer-b", false},
	} {
		revoked, err := policy.IsRevoked(candidate.kind, candidate.value, 1_800_000_000)
		if err != nil || revoked != candidate.expected {
			t.Fatalf("shared revocation decision mismatch: %v, %v", revoked, err)
		}
	}
}

func TestDurableRevocationEffectiveWindowRemoveAndReload(t *testing.T) {
	path := filepath.Join(t.TempDir(), "revocations.json")
	reader, _ := NewDurableRevocationPolicy(path, revocationOptions())
	writer, _ := NewDurableRevocationPolicy(path, revocationOptions())
	now := time.Now().Unix()
	if err := writer.Revoke(revocationEntry(t, RevocationPeerID, "future-peer", now, now+10, now+20)); err != nil {
		t.Fatal(err)
	}
	if revoked, _ := reader.IsRevoked(RevocationPeerID, "future-peer", now+11); revoked {
		t.Fatal("reader changed without explicit reload")
	}
	if err := reader.Reload(); err != nil {
		t.Fatal(err)
	}
	for timestamp, expected := range map[int64]bool{now + 9: false, now + 10: true, now + 20: false} {
		revoked, err := reader.IsRevoked(RevocationPeerID, "future-peer", timestamp)
		if err != nil || revoked != expected {
			t.Fatalf("unexpected decision at %d: %v, %v", timestamp, revoked, err)
		}
	}
	removed, err := writer.Remove(RevocationPeerID, "future-peer")
	if err != nil || !removed {
		t.Fatalf("remove failed: %v, %v", removed, err)
	}
	if err := reader.Reload(); err != nil {
		t.Fatal(err)
	}
	if revoked, _ := reader.IsRevoked(RevocationPeerID, "future-peer", now+11); revoked {
		t.Fatal("removed revocation remained active")
	}
}

func TestDurableRevocationCapacityAndCorruptionFailClosed(t *testing.T) {
	path := filepath.Join(t.TempDir(), "revocations.json")
	options := revocationOptions()
	options.MaxEntries = 1
	policy, err := NewDurableRevocationPolicy(path, options)
	if err != nil {
		t.Fatal(err)
	}
	now := time.Now().Unix()
	if err := policy.Revoke(revocationEntry(t, RevocationPeerID, "peer-a", now, 0, 0)); err != nil {
		t.Fatal(err)
	}
	err = policy.Revoke(revocationEntry(t, RevocationPeerID, "peer-b", now, 0, 0))
	requireCode(t, err, "revocation_state_capacity")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var value map[string]any
	if err := json.Unmarshal(raw, &value); err != nil {
		t.Fatal(err)
	}
	value["checksum"] = "sha256:00"
	raw, _ = json.Marshal(value)
	if err := os.WriteFile(path, raw, 0o600); err != nil {
		t.Fatal(err)
	}
	_, err = NewDurableRevocationPolicy(path, options)
	requireCode(t, err, "security_state_corrupt")
	if _, statErr := os.Stat(path); !errors.Is(statErr, os.ErrNotExist) {
		t.Fatalf("corrupt state was not quarantined: %v", statErr)
	}
}

func TestArtifactVerificationUsesDurableSignerRevocation(t *testing.T) {
	publicKey, privateKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatal(err)
	}
	credential := ArtifactSigningCredential{SignerIdentity: "producer-a", PublicKey: publicKey}
	signer := &ArtifactSigner{PrivateKey: privateKey, SignerIdentity: "producer-a"}
	artifact, err := signer.SignArtifact("artifact-a", []byte("verified payload"), time.Now().Unix())
	if err != nil {
		t.Fatal(err)
	}
	revocations, err := NewDurableRevocationPolicy(filepath.Join(t.TempDir(), "revocations.json"), revocationOptions())
	if err != nil {
		t.Fatal(err)
	}
	policy := NewArtifactTrustPolicy([]ArtifactSigningCredential{credential})
	policy.RevocationPolicy = revocations
	if err := VerifySignedArtifact([]byte("verified payload"), artifact, policy, time.Now().Unix()); err != nil {
		t.Fatal(err)
	}
	entry := revocationEntry(t, RevocationSignerFingerprint, credential.Fingerprint(), time.Now().Unix(), 0, 0)
	if err := revocations.Revoke(entry); err != nil {
		t.Fatal(err)
	}
	requireCode(t, VerifySignedArtifact([]byte("verified payload"), artifact, policy, time.Now().Unix()), "artifact_signer_revoked")
}
